components/style/restyle_hints.rs

```diff
@@ -25,34 +25,44 @@ use smallvec::SmallVec;
 use std::borrow::Borrow;
 use std::cell::Cell;
 use std::clone::Clone;
+use std::cmp;
 use stylist::SelectorMap;
 
+/// When the ElementState of an element (like IN_HOVER_STATE) changes,
+/// certain pseudo-classes (like :hover) may require us to restyle that
+/// element, its siblings, and/or its descendants. Similarly, when various
+/// attributes of an element change, we may also need to restyle things with
+/// id, class, and attribute selectors. Doing this conservatively is
+/// expensive, and so we use RestyleHints to short-circuit work we know is
+/// unnecessary.
+#[derive(Debug, Clone, PartialEq)]
+pub struct RestyleHint {
+    /// Depths at which selector matching must be re-run.
+    ///
+    /// The least significant bit indicates that selector matching must be re-run
+    /// for the element itself, the second least significant bit for the element's
+    /// children, the third its grandchildren, and so on.  If the most significant
+    /// bit it set, it indicates that that selector matching must be re-run for
+    /// elements at that depth and all of their descendants.
+    match_self: u8,
```

Comment:
Great idea about the newtype!

Regarding the name, WDYT about `match_under_self`?  (I want a name that I can adjust to have a similar field to represent depth to restyle under later siblings, e.g. `match_later_siblings` or `match_under_later_siblings`.)